File: programming-techniques/async-programming.js
/**
 * Asynchronous Programming - JavaScript
 * 
 * Non-blocking code execution using callbacks, promises, and async/await.
 */

// Callbacks
function fetchDataCallback(url, callback) {
    setTimeout(() => {
        callback(null, { data: `Data from ${url}` });
    }, 1000);
}

// Promises
function fetchDataPromise(url) {
    return new Promise((resolve, reject) => {
        setTimeout(() => {
            if (url) {
                resolve({ data: `Data from ${url}` });
            } else {
                reject(new Error('Invalid URL'));
            }
        }, 1000);
    });
}

// Async/Await
async function fetchDataAsync(url) {
    await new Promise(resolve => setTimeout(resolve, 1000));
    return { data: `Data from ${url}` };
}

// Parallel execution
async function fetchMultiple(urls) {
    const promises = urls.map(url => fetchDataAsync(url));
    return await Promise.all(promises);
}

// Race condition
async function fetchWithTimeout(url, timeout) {
    const fetchPromise = fetchDataAsync(url);
    const timeoutPromise = new Promise((_, reject) => 
        setTimeout(() => reject(new Error('Timeout')), timeout)
    );
    return await Promise.race([fetchPromise, timeoutPromise]);
}

// Example usage
console.log('=== Asynchronous Programming ===\n');

// Callback
fetchDataCallback('/api/users', (err, data) => {
    console.log('Callback result:', data);
});

// Promise
fetchDataPromise('/api/posts')
    .then(data => console.log('Promise result:', data))
    .catch(err => console.error('Error:', err));

// Async/Await
(async () => {
    const data = await fetchDataAsync('/api/comments');
    console.log('Async result:', data);
    
    // Parallel
    const results = await fetchMultiple(['/api/1', '/api/2', '/api/3']);
    console.log('Parallel results:', results);
})();

module.exports = { fetchDataCallback, fetchDataPromise, fetchDataAsync, fetchMultiple };

File: programming-techniques/async-programming.py
"""
Asynchronous Programming - Python

Non-blocking code execution using async/await and asyncio.
"""

import asyncio
from typing import List, Dict


# Async function
async def fetch_data_async(url: str) -> Dict:
    await asyncio.sleep(1)  # Simulate network delay
    return {'data': f'Data from {url}'}


# Parallel execution
async def fetch_multiple(urls: List[str]) -> List[Dict]:
    tasks = [fetch_data_async(url) for url in urls]
    return await asyncio.gather(*tasks)


# Timeout handling
async def fetch_with_timeout(url: str, timeout: float) -> Dict:
    try:
        return await asyncio.wait_for(fetch_data_async(url), timeout=timeout)
    except asyncio.TimeoutError:
        return {'error': 'Timeout'}


# Example usage
async def main():
    print("=== Asynchronous Programming ===\n")
    
    # Single async call
    data = await fetch_data_async('/api/users')
    print("Async result:", data)
    
    # Parallel execution
    results = await fetch_multiple(['/api/1', '/api/2', '/api/3'])
    print("Parallel results:", results)
    
    # With timeout
    result = await fetch_with_timeout('/api/data', 0.5)
    print("With timeout:", result)


if __name__ == "__main__":
    asyncio.run(main())

